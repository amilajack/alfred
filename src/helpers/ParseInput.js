// @flow
import flatten from 'lodash.flatten';
import uniq from 'uniq';
import findit from 'findit';
import { statAsync } from '../providers';

function findJsFiles(dir: string) {
  return new Promise(((resolve, reject) => {
    const files = [];
    findit(dir).on('file', (file) => {
      // only return files ending in .js
      if (/\.js$/.test(file)) {
        files.push(file);
      }
    }).on('end', () => {
      resolve(files);
    }).on('error', reject);
  }));
}

export default function ParseInput(files: Array<string>): Promise<Array<string>> {
  return Promise.all(files.map(file => statAsync(file).then(stat => (
    stat.isDirectory()
      ? findJsFiles(file)
      : [file]))))
    .then(flatten).then(uniq);
}
