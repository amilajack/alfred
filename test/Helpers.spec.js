// @flow
import path from 'path';
import ParseInput from '../src/helpers/ParseInput';

describe('Helpers', () => {
  const testBasePath = path.join(__dirname, '..', 'test', 'fixtures');

  describe('ParseInput', () => {
    it('should parse directories and files', async () => {
      expect(await ParseInput([testBasePath])).toMatchSnapshot();
      expect(
        await ParseInput([
          testBasePath,
          path.join(__dirname, '..', 'test'),
          path.join(__dirname, '..', 'src', 'alfred.js')
        ])
      ).toMatchSnapshot();
    });

    it('should fail on non-js files', async () => {
      const files = await ParseInput([
        path.join(testBasePath, 'basic-test.ts')
      ]);
      expect(files).toMatchSnapshot();
    });

    it('should filter duplicate files', async () => {
      const files = await ParseInput([
        path.join(testBasePath, 'basic-test.ts'),
        path.join(testBasePath, 'basic-test.ts'),
        path.join(testBasePath, 'class-test.ts'),
        path.join(testBasePath, 'class-test.ts')
      ]);
      expect(files).toMatchSnapshot();
    });

    it.skip('should fail on non-existent files', async () => {});
  });

  describe.skip('Validation', () => {});
});
