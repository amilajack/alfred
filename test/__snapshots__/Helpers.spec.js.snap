// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpers ParseInput should fail on non-js files 1`] = `
Array [
  "/Users/amila/Documents/Projects/alfred/test/fixtures/basic-test.ts",
]
`;

exports[`Helpers ParseInput should filter duplicate files 1`] = `
Array [
  "/Users/amila/Documents/Projects/alfred/test/fixtures/basic-test.ts",
  "/Users/amila/Documents/Projects/alfred/test/fixtures/class-test.ts",
]
`;

exports[`Helpers ParseInput should parse directories and files 1`] = `
Array [
  "/Users/amila/Documents/Projects/alfred/test/fixtures/js-file-test.js",
  "/Users/amila/Documents/Projects/alfred/test/fixtures/test-folder/test-js.js",
]
`;

exports[`Helpers ParseInput should parse directories and files 2`] = `
Array [
  "/Users/amila/Documents/Projects/alfred/src/alfred.js",
  "/Users/amila/Documents/Projects/alfred/test/Helpers.spec.js",
  "/Users/amila/Documents/Projects/alfred/test/Migrate.spec.js",
  "/Users/amila/Documents/Projects/alfred/test/fixtures/js-file-test.js",
  "/Users/amila/Documents/Projects/alfred/test/fixtures/test-folder/test-js.js",
]
`;
